Userland/Libraries/LibAudio/ConnectionFromClient.cpp

```diff
@@ -1,48 +1,126 @@
+    , m_user_queue(make<UserSampleQueue>())
+    , m_background_audio_enqueuer(Threading::Thread::construct([this]() {
+        // All the background thread does is run an event loop
+        Core::EventLoop enqueuer_loop;
+        m_enqueuer_loop = &enqueuer_loop;
+        enqueuer_loop.exec();
+        m_enqueuer_loop = nullptr;
+        return (intptr_t) nullptr;
+    }))
+{
+    m_background_audio_enqueuer->start();
+    set_buffer(*m_buffer);
+}
+
+ConnectionFromClient::~ConnectionFromClient()
 {
+    die();
 }
 
-void ConnectionFromClient::enqueue(Buffer const& buffer)
+void ConnectionFromClient::die()
 {
-    for (;;) {
-        auto success = enqueue_buffer(buffer.anonymous_buffer(), buffer.id(), buffer.sample_count());
-        if (success)
+    // We're sometimes getting here after the other thread has already exited and its event loop does no longer exist.
+    if (m_enqueuer_loop != nullptr) {
+        m_enqueuer_loop->wake();
+        m_enqueuer_loop->quit(0);
+    }
+    (void)m_background_audio_enqueuer->join();
+}
+
+ErrorOr<void> ConnectionFromClient::async_enqueue(FixedArray<Sample>&& samples)
+{
+    update_good_sleep_time();
+    m_user_queue->append(move(samples));
+    // Wake the background thread to make sure it starts enqueuing audio.
+    if (!m_audio_enqueuer_active.load())
+        m_enqueuer_loop->post_event(*this, make<Core::CustomEvent>(0), Core::EventLoop::ShouldWake::Yes);
+    async_start_playback();
+
+    return {};
+}
+
+void ConnectionFromClient::clear_client_buffer()
+{
+    m_user_queue->clear();
+}
+
+void ConnectionFromClient::update_good_sleep_time()
+{
+    auto sample_rate = static_cast<double>(get_sample_rate());
+    auto buffer_play_time_ns = 1'000'000'000.0 / (sample_rate / static_cast<double>(AUDIO_BUFFER_SIZE));
+    // A factor of 1 should be good for now.
+    m_good_sleep_time = Time::from_nanoseconds(static_cast<unsigned>(buffer_play_time_ns)).to_timespec();
+}
+
+// Non-realtime audio writing loop
+void ConnectionFromClient::custom_event(Core::CustomEvent&)
+{
+    while (true) {
+        m_audio_enqueuer_active.store(true);
+        Array<Sample, AUDIO_BUFFER_SIZE> next_chunk;
+
+        if (m_user_queue->is_empty()) {
+            dbgln("Reached end of provided audio data, going to sleep");
             break;
-        nanosleep(&g_enqueue_wait_time, nullptr);
+        }
+
+        auto available_samples = min(AUDIO_BUFFER_SIZE, m_user_queue->size());
+        for (size_t i = 0; i < available_samples; ++i) {
+            next_chunk[i] = (*m_user_queue)[i];
+        }
+
+        m_user_queue->slice(available_samples);
+
+        // FIXME: Could we recieve interrupts in a good non-IPC way instead?
+        auto result = m_buffer->try_blocking_enqueue(next_chunk, [this]() {
+            nanosleep(&m_good_sleep_time, nullptr);
+        });
+        if (result.is_error())
+            dbgln("Error while writing samples to shared buffer: {}", result.error());
```

Comment:
Looking at the underlying code, we can't ever return anything except a full queue error. So that's perfectly fine to ignore and try again. The audio server also does this, but in reverse.